kernel/src/client/arrow_utils.rs

```diff
@@ -2,39 +2,60 @@
 
 use arrow_array::RecordBatch;
 use arrow_schema::{Schema as ArrowSchema, SchemaRef as ArrowSchemaRef};
-use itertools::Itertools;
 use parquet::{arrow::ProjectionMask, schema::types::SchemaDescriptor};
 
-/// Get the indicies in `parquet_schema` of the specified columns in `requested_schema`
+/// Get the indicies in `parquet_schema` of the specified columns in `requested_schema`. This
+/// returns a tuples of (mask_indicies: Vec<parquet_schema_index>, reorder_indicies:
+/// Vec<requested_index>). `mask_indicies` is used for generating the mask for reading from the
+/// parquet file, and simply contains an entry for each index we wish to select from the parquet
+/// file set to the index of the requested column in the parquet. `reorder_indicies` is used for
+/// re-ordering and will be the same size as `requested_schema`. Each index in `reorder_indicies`
+/// represents a column that will be in the read parquet data at that index. The value stored in
+/// `reorder_indicies` is the position that the column should appear in the final output. For
+/// example, if `reorder_indicies` is `[2,0,1]`, then the re-ordering code should take the third
+/// column in the raw-read parquet data, and move it to the first column in the final output, the
+/// first column to the second, and the second to the third.
 pub(crate) fn get_requested_indices(
-    requested_schema: &ArrowSchema,
+    requested_schema: &SchemaRef,
     parquet_schema: &ArrowSchemaRef,
-) -> DeltaResult<Vec<usize>> {
-    let indicies = requested_schema
-        .fields
+) -> DeltaResult<(Vec<usize>, Vec<usize>)> {
+    let requested_len = requested_schema.fields.len();
+    let mut mask_indicies = vec![0; requested_len];
+    let mut found_count = 0; // verify that we found all requested fields
+    let reorder_indicies = parquet_schema
+        .fields()
         .iter()
-        .map(|field| {
-            // todo: handle nested (then use `leaves` not `roots` below in generate_mask)
-            parquet_schema.index_of(field.name())
+        .enumerate()
+        .filter_map(|(parquet_position, field)| {
+            requested_schema.index_of(field.name()).map(|index| {
+                found_count += 1;
+                mask_indicies[index] = parquet_position;
+                index
+            })
         })
-        .try_collect()?;
-    Ok(indicies)
+        .collect();
+    if found_count != requested_len {
+        return Err(Error::generic(
+            "Didn't find all requested columns in parquet schema",
+        ));
+    }
+    Ok((mask_indicies, reorder_indicies))
 }
 
 /// Create a mask that will only select the specified indicies from the parquet. Currently we only
 /// handle "root" level columns, and hence use `ProjectionMask::roots`, but will support leaf
 /// selection in the future. See issues #86 and #96 as well.
 pub(crate) fn generate_mask(
-    requested_schema: &ArrowSchema,
+    requested_schema: &SchemaRef,
     parquet_schema: &ArrowSchemaRef,
```

Comment:
What purpose does the `parquet_schema` arg serve, as different from the `parquet_physical_schema`? Seems like we could just check the latter's length at L58 below, and then wouldn't need the extra arg any more?